File: src/main/java/xyz/oribuin/playerwarps/manager/ConfigurationManager.java
package xyz.oribuin.playerwarps.manager;


import dev.rosewood.rosegarden.RosePlugin;
import dev.rosewood.rosegarden.config.CommentedFileConfiguration;
import dev.rosewood.rosegarden.config.RoseSetting;
import dev.rosewood.rosegarden.manager.AbstractConfigurationManager;
import xyz.oribuin.playerwarps.PlayerWarpsPlugin;

import java.util.List;

public class ConfigurationManager extends AbstractConfigurationManager {

    public ConfigurationManager(RosePlugin rosePlugin) {
        super(rosePlugin, Setting.class);
    }

    @Override
    protected String[] getHeader() {
        return new String[]{};
    }

    public enum Setting implements RoseSetting {
        DISABLED_WORLDS("disabled-worlds", List.of("disabled-world-1"), "Here you can configure the worlds that warps are disabled in."),
        DATE_FORMAT("date-format", "dd/MMM/yy", "The date format to use for dates in the plugin."),

        WARP_SETTINGS("warp-settings", null, "Here you can configure the settings for warps."),
        WARP_CREATE_COST("warp-settings.create-cost", 500.0, "The cost to create a warp.", "Uses vault as the economy provider."),
        WARP_DELETE_COST("warp-settings.delete-cost", 500.0, "The cost to delete a warp.", "Uses vault as the economy provider."),
        WARP_DESC_FILTERS("warp-settings.desc-filters", List.of("\\b(f+(\\W|\\d|_)*(a|@)+(\\W|\\d|_)*g+(\\W|\\d|_)*)\\b"), "If any of these patterns are matched, the warp description will be invalid."),
        WARP_NAME_FILTERS("warp-settings.name-filters", List.of("\\b(f+(\\W|\\d|_)*(a|@)+(\\W|\\d|_)*g+(\\W|\\d|_)*)\\b"), "If any of these patterns are matched, the warp name will be invalid."),

        // Upkeep Settings
        WARP_UPKEEP_ENABLED("warp-settings.upkeep-enabled", false, "If true, warps will require a regular payment to stay active."),
        WARP_UPKEEP_COST("warp-settings.upkeep-cost", 500.0, "The cost to keep a warp active.", "Uses vault as the economy provider."),
        WARP_UPKEEP_INTERVAL("warp-settings.upkeep-interval", 604800, "The interval in seconds to charge the upkeep cost."),
        WARP_UPKEEP_WARN("warp-settings.upkeep-warn", true, "If true, The owner of the warp will be warned a minute before the upkeep is charged."),

        ;

        private final String key;
        private final Object defaultValue;
        private final String[] comments;
        private Object value = null;

        Setting(String key, Object defaultValue, String... comments) {
            this.key = key;
            this.defaultValue = defaultValue;
            this.comments = comments != null ? comments : new String[0];
        }

        @Override
        public String getKey() {
            return this.key;
        }

        @Override
        public Object getDefaultValue() {
            return this.defaultValue;
        }

        @Override
        public String[] getComments() {
            return this.comments;
        }

        @Override
        public Object getCachedValue() {
            return this.value;
        }

        @Override
        public void setCachedValue(Object value) {
            this.value = value;
        }

        @Override
        public CommentedFileConfiguration getBaseConfig() {
            return PlayerWarpsPlugin.get().getManager(ConfigurationManager.class).getConfig();
        }
    }
}

File: src/main/kotlin/xyz/oribuin/playerwarps/manager/ConfigurationManager.kt
package xyz.oribuin.playerwarps.manager


import dev.rosewood.rosegarden.RosePlugin
import dev.rosewood.rosegarden.config.CommentedFileConfiguration
import dev.rosewood.rosegarden.config.RoseSetting
import dev.rosewood.rosegarden.manager.AbstractConfigurationManager
import org.bukkit.Material
import xyz.oribuin.playerwarps.PlayerWarpsPlugin
import xyz.oribuin.playerwarps.util.WarpUtils.getManager

class ConfigurationManager(rosePlugin: RosePlugin) : AbstractConfigurationManager(rosePlugin, Setting::class.java) {

    enum class Setting(private val key: String?, private val defaultValue: Any?, private vararg val comments: String?) :
        RoseSetting {
        // General Warp Settings
        DISABLED_WORLDS("disabled-worlds", emptyList<String>(), "A list of worlds where warps are disabled."),
        DISABLED_ICONS("disabled-icons", listOf(Material.BEDROCK.name), "A list of icons that are disabled."),
        MIN_WARP_LIMIT("min-warp-limit", 1, "The minimum amount of warps a player can have."),
        DATE_FORMAT("date-format", "dd/MMMM", "What format the date should be displayed in."),

        // Command Cooldown Settings
        COOLDOWNS("cooldowns", null, "The cooldowns for each command.", "Set each option as 0 to disable."),
        COOLDOWNS_CREATE("cooldowns.create", 60, "The cooldown (in seconds) for creating a warp."),
        COOLDOWNS_DELETE("cooldowns.delete", 120, "The cooldown (in seconds) for deleting a warp."),
        COOLDOWNS_TELEPORT("cooldowns.teleport", 10, "The cooldown (in seconds) for teleporting to a warp."),
        COOLDOWNS_RENAME("cooldowns.rename", 60, "The cooldown (in seconds) for renaming a warp."),

        // Warp Pricing Settings
        ECONOMY_PLUGIN("economy.provider", "VAULT", "The primary economy plugin to use for all economy related features.", "Options: VAULT, PLAYERPOINTS, TOKENMANAGER, TREASURY"),
        CREATE_WARP_PRICE("economy.create", 1000.0, "The cost for creating a warp."),
        DELETE_WARP_PRICE("economy.delete", 1000.0, "The cost for deleting a warp."),
        RENAME_WARP_PRICE("economy.rename", 1000.0, "The cost for renaming a warp."),
        MIN_TELEPORT_PRICE("economy.min-teleport", 5.0, "The minimum cost a player can set for teleporting to their warp."),
        MAX_TELEPORT_PRICE("economy.max-teleport", 100.0, "The maximum cost a player can set for teleporting to their warp."),

        //
        ; // TODO: Add settings

        private var value: Any? = null

        override fun getKey() = this.key

        override fun getDefaultValue() = this.defaultValue

        override fun getComments() = this.comments

        override fun getCachedValue() = this.value

        override fun setCachedValue(value: Any?) {
            this.value = value
        }

        override fun getBaseConfig(): CommentedFileConfiguration =
            PlayerWarpsPlugin.instance.getManager<ConfigurationManager>().config

    }

    //__________.__                             __      __
    //\______   \  | _____  ___.__. ___________/  \    /  \_____ _____________  ______
    // |     ___/  | \__  \<   |  |/ __ \_  __ \   \/\/   /\__  \\_  __ \____ \/  ___/
    // |    |   |  |__/ __ \\___  \  ___/|  | \/\        /  / __ \|  | \/  |_> >___ \
    // |____|   |____(____  / ____|\___  >__|    \__/\  /  (____  /__|  |   __/____  >
    //                    \/\/         \/             \/        \/      |__|       \/
    override fun getHeader(): Array<String> {
        return arrayOf(
            "__________.__                             __      __                            ",
            "\\______   \\  | _____  ___.__. ___________/  \\    /  \\_____ _____________  ______",
            " |     ___/  | \\__  \\<   |  |/ __ \\_  __ \\   \\/\\/   /\\__  \\\\_  __ \\____ \\/  ___/",
            " |    |   |  |__/ __ \\\\___  \\  ___/|  | \\/\\        /  / __ \\|  | \\/  |_> >___ \\ ",
            " |____|   |____(____  / ____|\\___  >__|    \\__/\\  /  (____  /__|  |   __/____  >",
            "                    \\/\\/         \\/             \\/        \\/      |__|       \\/ "
        )
    }
}
